ballista/rust/scheduler/src/state/task_manager.rs

```diff
@@ -608,29 +602,91 @@ impl<T: 'static + AsLogicalPlan, U: 'static + AsExecutionPlan> TaskManager<T, U>
             .collect()
     }
 
-    pub fn increase_pending_queue_size(&self, num: usize) {
-        let result = self
-            .pending_task_queue_size
-            .fetch_add(num, AOrdering::Relaxed);
-        self.metrics_collector
-            .set_pending_tasks_queue_size(result as f64);
+    pub fn increase_pending_queue_size(&self, num: usize) -> Result<()> {
+        if num != 0 {
+            loop {
+                let old_value = self.get_pending_task_queue_size();
+
+                if usize::MAX - old_value >= num {
+                    let new_value = old_value + num;
+                    match self.pending_task_queue_size.compare_exchange(
+                        old_value,
+                        new_value,
+                        AOrdering::AcqRel,
+                        AOrdering::Relaxed,
+                    ) {
+                        Ok(_) => {
+                            self.metrics_collector
+                                .set_pending_tasks_queue_size(new_value as f64);
+                            debug!(
+                                "Pending queue size {} increased by {} to {}",
+                                old_value, num, new_value
+                            );
+                            break;
+                        }
+                        Err(_) => {
+                            error!(
+                                "Unable to increase pending queue size {} by {} to {}",
+                                old_value, num, new_value
+                            );
+                            continue;
+                        }
+                    }
+                } else {
+                    return Err(BallistaError::Internal(format!(
+                        "Refused to increase pending queue size {} by {}, wrap on overflow will happen", 
+                        old_value, num
+                    )));
```

Comment:
I don't think we should fail the call, only log an error. This should not prevent new tasks from being scheduled.